File: platform/services/processing/stream/feature-engine/src/engine/indicators.rs
use std::collections::BTreeMap;

use crate::config::IndicatorConfig;

/// Dynamic feature map — keys are indicator names (e.g., "rolling_mean_7"),
/// values are computed f64 values. Use-cases control which indicators appear
/// by configuring `IndicatorConfig` in their config file.
pub type Features = BTreeMap<String, f64>;

/// Compute all configured indicators from primary/secondary value time series.
/// Which indicators are computed and with what periods is driven by config,
/// not hardcoded. These are generic time series operations (windowed averages,
/// momentum, dispersion) applicable to any numeric data — not domain-specific.
/// Use-cases provide their own indicator configuration via config file.
pub fn compute_all(
    primary_values: &[f64],
    secondary_values: &[f64],
    config: &IndicatorConfig,
) -> Features {
    let mut features = Features::new();

    if primary_values.is_empty() {
        return features;
    }

    features.insert("value".to_string(), *primary_values.last().unwrap_or(&0.0));

    // Rolling mean — configurable periods
    for &period in &config.rolling_mean_periods {
        features.insert(
            format!("rolling_mean_{}", period),
            compute_rolling_mean(primary_values, period),
        );
    }

    // Exponential rolling mean — configurable periods
    for &period in &config.exp_avg_periods {
        features.insert(
            format!("rolling_ema_{}", period),
            compute_exp_avg(primary_values, period),
        );
    }

    // Momentum oscillator — configurable period
    features.insert(
        format!("momentum_{}", config.momentum_period),
        compute_momentum(primary_values, config.momentum_period),
    );

    // Trend convergence — optional
    if config.trend_convergence_enabled {
        let (convergence, signal, hist) = compute_trend_convergence(primary_values);
        features.insert("trend_convergence".to_string(), convergence);
        features.insert("trend_signal".to_string(), signal);
        features.insert("trend_hist".to_string(), hist);
    }

    // Deviation bands — optional, configurable period
    if config.deviation_bands_enabled {
        let (upper, middle, lower) =
            compute_deviation_bands(primary_values, config.deviation_bands_period);
        features.insert("band_upper".to_string(), upper);
        features.insert("band_middle".to_string(), middle);
        features.insert("band_lower".to_string(), lower);
    }

    // Secondary value rolling mean
    if !secondary_values.is_empty() {
        features.insert(
            "secondary_avg".to_string(),
            compute_rolling_mean(secondary_values, config.secondary_avg_period),
        );
    }

    // Value change and dispersion
    features.insert(
        "value_change".to_string(),
        compute_value_change(primary_values),
    );
    features.insert(
        "dispersion".to_string(),
        compute_dispersion(primary_values, config.dispersion_period),
    );

    features
}

pub fn compute_rolling_mean(values: &[f64], period: usize) -> f64 {
    if values.len() < period {
        return 0.0;
    }
    let slice = &values[values.len() - period..];
    slice.iter().sum::<f64>() / period as f64
}

pub fn compute_exp_avg(values: &[f64], period: usize) -> f64 {
    if values.is_empty() {
        return 0.0;
    }

    let k = 2.0 / (period as f64 + 1.0);
    let mut result = values[0];

    for val in values.iter().skip(1) {
        result = val * k + result * (1.0 - k);
    }

    result
}

pub fn compute_momentum(values: &[f64], period: usize) -> f64 {
    if values.len() < period + 1 {
        return 50.0;
    }

    let mut gains = 0.0;
    let mut losses = 0.0;

    for i in (values.len() - period)..values.len() {
        let change = values[i] - values[i - 1];
        if change > 0.0 {
            gains += change;
        } else {
            losses -= change;
        }
    }

    let avg_gain = gains / period as f64;
    let avg_loss = losses / period as f64;

    if avg_loss == 0.0 {
        return 100.0;
    }

    let rs = avg_gain / avg_loss;
    100.0 - (100.0 / (1.0 + rs))
}

pub fn compute_trend_convergence(values: &[f64]) -> (f64, f64, f64) {
    if values.len() < 26 {
        return (0.0, 0.0, 0.0);
    }

    let fast = compute_exp_avg(values, 12);
    let slow = compute_exp_avg(values, 26);
    let convergence = fast - slow;

    let signal = convergence * (2.0 / 10.0);
    let hist = convergence - signal;

    (convergence, signal, hist)
}

pub fn compute_deviation_bands(values: &[f64], period: usize) -> (f64, f64, f64) {
    if values.len() < period {
        return (0.0, 0.0, 0.0);
    }

    let slice = &values[values.len() - period..];
    let mean = slice.iter().sum::<f64>() / period as f64;

    let variance = slice.iter().map(|v| (v - mean).powi(2)).sum::<f64>() / period as f64;
    let std_dev = variance.sqrt();

    let upper = mean + 2.0 * std_dev;
    let lower = mean - 2.0 * std_dev;

    (upper, mean, lower)
}

pub fn compute_value_change(values: &[f64]) -> f64 {
    if values.len() < 2 {
        return 0.0;
    }
    let current = values[values.len() - 1];
    let previous = values[values.len() - 2];
    if previous == 0.0 {
        0.0
    } else {
        (current - previous) / previous * 100.0
    }
}

pub fn compute_dispersion(values: &[f64], period: usize) -> f64 {
    if values.len() < period {
        return 0.0;
    }

    let slice = &values[values.len() - period..];
    let mean = slice.iter().sum::<f64>() / period as f64;
    let variance = slice.iter().map(|v| (v - mean).powi(2)).sum::<f64>() / period as f64;
    variance.sqrt()
}

File: platform/services/processing/stream/src/main/java/pipeline/functions/Indicators.java
package pipeline.functions;

import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * Pure, stateless time-series indicator math - the Java counterpart of the Rust
 * feature-engine {@code indicators.rs}. Every formula here is a faithful port of
 * its Rust sibling so the Flink stream processor and the (retired) Rust engine
 * are byte-for-byte interchangeable on the same input window. Kept free of any
 * Flink / Jackson types so it is unit-testable in isolation (see IndicatorsTest)
 * - those tests ARE the no-lost-service proof for the feature-engine to Flink
 * cutover.
 *
 * <p>All operations are generic windowed statistics (means, EMA, momentum,
 * dispersion) applicable to any numeric series - no domain logic. The MACD-style
 * <em>signal line</em> is intentionally NOT here: it requires per-key state
 * (an EMA over the convergence series itself) and therefore lives in
 * {@link FeatureFunction}.
 */
public final class Indicators {
    private Indicators() {}

    /** Simple moving average of the last {@code period} values; 0 if too short. */
    public static double rollingMean(List<Double> values, int period) {
        if (period <= 0 || values.size() < period) {
            return 0.0;
        }
        double sum = 0.0;
        for (int i = values.size() - period; i < values.size(); i++) {
            sum += values.get(i);
        }
        return sum / period;
    }

    /**
     * Exponential moving average over the full series, seeded at the first
     * element. Mirrors Rust {@code compute_exp_avg}: returns 0 only when EMPTY
     * (NOT when shorter than {@code period}); otherwise smooths from values[0].
     */
    public static double expAvg(List<Double> values, int period) {
        if (values.isEmpty()) {
            return 0.0;
        }
        double k = 2.0 / (period + 1.0);
        double result = values.get(0);
        for (int i = 1; i < values.size(); i++) {
            result = values.get(i) * k + result * (1.0 - k);
        }
        return result;
    }

    /**
     * RSI-style momentum oscillator over the last {@code period} deltas.
     * Mirrors Rust {@code compute_momentum}: 50 when fewer than period+1
     * points, 100 when there are no losses.
     */
    public static double momentum(List<Double> values, int period) {
        if (values.size() < period + 1) {
            return 50.0;
        }
        double gains = 0.0;
        double losses = 0.0;
        for (int i = values.size() - period; i < values.size(); i++) {
            double change = values.get(i) - values.get(i - 1);
            if (change > 0.0) {
                gains += change;
            } else {
                losses -= change;
            }
        }
        double avgGain = gains / period;
        double avgLoss = losses / period;
        if (avgLoss == 0.0) {
            return 100.0;
        }
        double rs = avgGain / avgLoss;
        return 100.0 - (100.0 / (1.0 + rs));
    }

    /**
     * Deviation bands (mean ± 2·σ) over the last {@code period} values, where σ
     * is the POPULATION standard deviation. Mirrors Rust
     * {@code compute_deviation_bands}. Returns {@code [upper, middle, lower]};
     * all zero when fewer than {@code period} points.
     */
    public static double[] deviationBands(List<Double> values, int period) {
        if (period <= 0 || values.size() < period) {
            return new double[] {0.0, 0.0, 0.0};
        }
        double mean = rollingMean(values, period);
        double variance = 0.0;
        for (int i = values.size() - period; i < values.size(); i++) {
            double d = values.get(i) - mean;
            variance += d * d;
        }
        variance /= period;
        double std = Math.sqrt(variance);
        return new double[] {mean + 2.0 * std, mean, mean - 2.0 * std};
    }

    /**
     * POPULATION standard deviation over the last {@code period} values.
     * Mirrors Rust {@code compute_dispersion} (NOT a sample/whole-window std);
     * 0 when fewer than {@code period} points.
     */
    public static double dispersion(List<Double> values, int period) {
        if (period <= 0 || values.size() < period) {
            return 0.0;
        }
        double mean = rollingMean(values, period);
        double variance = 0.0;
        for (int i = values.size() - period; i < values.size(); i++) {
            double d = values.get(i) - mean;
            variance += d * d;
        }
        variance /= period;
        return Math.sqrt(variance);
    }

    /**
     * Percent change between the last two values. Mirrors Rust
     * {@code compute_value_change}: ×100 (percent), 0 when fewer than two
     * points or the previous value is zero.
     */
    public static double valueChange(List<Double> values) {
        if (values.size() < 2) {
            return 0.0;
        }
        double current = values.get(values.size() - 1);
        double previous = values.get(values.size() - 2);
        if (previous == 0.0) {
            return 0.0;
        }
        return (current - previous) / previous * 100.0;
    }

    /**
     * Compute every configured STATELESS indicator from the primary/secondary
     * series, mirroring Rust {@code compute_all}. Returns a sorted map (matches
     * the serde {@code BTreeMap} key ordering of the feature-engine cache).
     *
     * <p>The trend (MACD) triple is added by {@link FeatureFunction} because its
     * signal line is stateful; everything else is self-contained here.
     */
    public static Map<String, Double> computeAll(
            List<Double> primaryValues, List<Double> secondaryValues, IndicatorConfig config) {
        Map<String, Double> features = new TreeMap<>();
        if (primaryValues.isEmpty()) {
            return features;
        }

        features.put("value", primaryValues.get(primaryValues.size() - 1));

        for (int period : config.rollingMeanPeriods) {
            features.put("rolling_mean_" + period, rollingMean(primaryValues, period));
        }
        for (int period : config.expAvgPeriods) {
            features.put("rolling_ema_" + period, expAvg(primaryValues, period));
        }

        features.put("momentum_" + config.momentumPeriod, momentum(primaryValues, config.momentumPeriod));

        if (config.deviationBandsEnabled) {
            double[] bands = deviationBands(primaryValues, config.deviationBandsPeriod);
            features.put("band_upper", bands[0]);
            features.put("band_middle", bands[1]);
            features.put("band_lower", bands[2]);
        }

        if (!secondaryValues.isEmpty()) {
            features.put("secondary_avg", rollingMean(secondaryValues, config.secondaryAvgPeriod));
        }

        features.put("value_change", valueChange(primaryValues));
        features.put("dispersion", dispersion(primaryValues, config.dispersionPeriod));

        return features;
    }
}
